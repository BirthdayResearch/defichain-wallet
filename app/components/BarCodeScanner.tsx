import React, { useState, useEffect } from 'react'
import { Text } from 'react-native'
import { BarCodeScanner as DefaultBarCodeScanner } from 'expo-barcode-scanner'
import { Logging } from '../api/logging'
import { translate } from '../translations'
import tailwind from 'tailwind-rn'
import { View } from '.'
import { StackScreenProps } from '@react-navigation/stack'
import { BalanceParamList } from '../screens/AppNavigator/screens/Balances/BalancesNavigator'

type Props = StackScreenProps<BalanceParamList, 'BarCodeScanner'>

export function BarCodeScanner ({ route, navigation }: Props): JSX.Element {
  // null => undetermined
  const [hasPermission, setHasPermission] = useState<boolean | null>(null)

  useEffect(() => {
    DefaultBarCodeScanner.requestPermissionsAsync()
      .then(({ status }) => {
        switch (status) {
          case 'granted': setHasPermission(true); break
          case 'denied': setHasPermission(false); break
        }
      })
      .catch(e => Logging.error(e))
  }, [])

  const handleBarCodeScanned = ({ data }: { type: string, data: string }): void => {
    route.params.onQrScanned(data)
    navigation.goBack()
  }

  if (hasPermission === null) {
    return (
      <View style={tailwind('flex-col flex-1 justify-center items-center')}>
        <Text>{translate('components/BarCodeScanner', 'Requesting for camera permission')}</Text>
      </View>
    )
  }
  if (!hasPermission) {
    <View style={tailwind('flex-col flex-1 justify-center items-center')}>
      <Text>{translate('components/BarCodeScanner', 'You have denied the permission request to use your camera')}</Text>
    </View>
  }

  return (
    <DefaultBarCodeScanner
      style={tailwind('flex-1')}
      onBarCodeScanned={handleBarCodeScanned}
    />
  )
}
