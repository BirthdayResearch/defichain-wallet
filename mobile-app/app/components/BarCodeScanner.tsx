import { StackScreenProps } from '@react-navigation/stack'
import { BarCodeScanner as DefaultBarCodeScanner } from 'expo-barcode-scanner'
import { useEffect, useState } from 'react'
import { StyleSheet } from 'react-native'
import tailwind from 'tailwind-rn'
import { View } from '.'
import { BalanceParamList } from '@screens/AppNavigator/screens/Balances/BalancesNavigator'
import { translate } from '@translations'
import { ThemedText } from './themed'
import { useLogger } from '@shared-contexts/NativeLoggingProvider'

type Props = StackScreenProps<BalanceParamList, 'BarCodeScanner'>

export function BarCodeScanner ({
  route,
  navigation
}: Props): JSX.Element {
  // null => undetermined
  const [hasPermission, setHasPermission] = useState<boolean | null>(null)
  const [value, setValue] = useState<string>()
  const logger = useLogger()

  // to ensure callback only can be fired once as value change
  useEffect(() => {
    if (value !== undefined) {
      route.params.onQrScanned(value)
      navigation.pop()
    }
  }, [value])

  useEffect(() => {
    DefaultBarCodeScanner.requestPermissionsAsync()
      .then(({ status }) => {
        switch (status) {
          case 'granted':
            setHasPermission(true)
            break
          case 'denied':
            setHasPermission(false)
            break
        }
      })
      .catch(logger.error)
  }, [])

  if (hasPermission === null) {
    return (
      <View style={tailwind('flex-col flex-1 justify-center items-center')}>
        <ThemedText>
          {translate('components/BarCodeScanner', 'Requesting for camera permission')}
        </ThemedText>
      </View>
    )
  }
  if (!hasPermission) {
    return (
      <View style={tailwind('flex-col flex-1 justify-center items-center')}>
        <ThemedText>
          {translate('components/BarCodeScanner', 'You have denied the permission request to use your camera')}
        </ThemedText>
      </View>
    )
  }

  const opacity = 'rgba(0, 0, 0, .6)'
  const styles = StyleSheet.create({
    focused: {
      flex: 5
    },
    layerBottom: {
      backgroundColor: opacity,
      flex: 2
    },
    layerCenter: {
      flex: 3,
      // aspectRatio: 1,
      flexDirection: 'row'
    },
    layerLeft: {
      backgroundColor: opacity,
      flex: 1
    },
    layerRight: {
      backgroundColor: opacity,
      flex: 1
    },
    layerTop: {
      backgroundColor: opacity,
      flex: 2
    }
  })

  return (
    <>
      <DefaultBarCodeScanner
        barCodeTypes={[DefaultBarCodeScanner.Constants.BarCodeType.qr]}
        onBarCodeScanned={(e) => {
          setValue(e.data)
        }}
        style={tailwind('flex-1 absolute inset-0')}
      />

      <View style={styles.layerTop} />

      <View style={styles.layerCenter}>
        <View style={styles.layerLeft} />

        <View style={[styles.focused, tailwind('relative')]}>
          <View style={tailwind('border-t-4 border-l-4 border-white w-16 h-16 absolute')} />

          <View style={tailwind('border-t-4 border-r-4 border-white w-16 h-16 top-0 right-0 absolute')} />

          <View style={tailwind('border-b-4 border-l-4 border-white w-16 h-16 bottom-0 absolute')} />

          <View style={tailwind('border-b-4 border-r-4 border-white w-16 h-16 bottom-0 right-0 absolute')} />
        </View>

        <View style={styles.layerRight} />
      </View>

      <View style={styles.layerBottom} />
    </>
  )
}
