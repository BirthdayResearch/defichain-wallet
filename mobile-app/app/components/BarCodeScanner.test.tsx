import { render } from '@testing-library/react-native'
import { BarCodeScanner } from './BarCodeScanner'

jest.mock('@shared-contexts/ThemeProvider')

describe('barcode scanner', () => {
  it('should match snapshot', async () => {
    const navigation: any = {
      navigate: jest.fn()
    }
    const route: any = {}
    const component = (
      <BarCodeScanner
        navigation={navigation}
        route={route}
      />
    )
    const rendered = render(component)
    expect(rendered.toJSON()).toMatchSnapshot()
  })
})
