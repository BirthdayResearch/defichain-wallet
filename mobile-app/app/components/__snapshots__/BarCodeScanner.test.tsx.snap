// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`barcode scanner should match snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexBasis": "0%",
      "flexDirection": "column",
      "flexGrow": 1,
      "flexShrink": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "RegularFont",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 24,
        },
        Array [
          undefined,
          Object {
            "color": "rgba(0, 0, 0, 1)",
          },
        ],
      ]
    }
  >
    Requesting for camera permission
  </Text>
</View>
`;
